pedsnetdcc/primary_keys.py

```diff
@@ -0,0 +1,130 @@
+import logging
+import re
+import sqlalchemy
+import sqlalchemy.dialects.postgresql
+import time
+
+from psycopg2 import errorcodes as psycopg2_errorcodes
+
+from pedsnetdcc import VOCAB_TABLES
+from pedsnetdcc.db import StatementSet, Statement
+from pedsnetdcc.dict_logging import secs_since
+from pedsnetdcc.utils import (stock_metadata, combine_dicts,
+                              get_conn_info_dict)
+
+
+logger = logging.getLogger(__name__)
+
+
+def _primary_keys_from_model_version(model_version):
+    """Return list of SQLAlchemy primary key objects for the model version.
+
+    Relevant attributes are `name` and `table.name`
+
+    :param str model_version: pedsnet model version
+    :return:                  list of primary keys objects
+    :rtype:                   list(sqlalchemy.PrimaryKey)
+    """
+
+    # TODO: should ideally be based on the transformed data model
+    # in case a transform modifies a primary key.
+    metadata = stock_metadata(model_version)
+
+    primary_keys = []
+    for name, table in metadata.tables.items():
+        if name not in VOCAB_TABLES:
+            for con in table.constraints:
+                if con and isinstance(con, sqlalchemy.PrimaryKeyConstraint):
+                    primary_keys.append(con)
+
+    return primary_keys
+
+
+def _check_stmt_err(stmt, force):
+    """Validate results of an executed Statement object relative to force flag
+
+    Creating a constraint can produce a 'benign' error if it already exists.
+    If `force` is true, such an error is ignored.
+
+    :param Statement stmt:
+    :param bool force:
+    :return: None
+    :raise: DatabaseError
+    """
+    if stmt.err is None:
+        return
+
+    # Detect error 42P16: multiple primary keys ... are not allowed
+    already_exists = (
+        hasattr(stmt.err, 'pgcode')
+        and stmt.err.pgcode
+        and psycopg2_errorcodes.lookup(
+            stmt.err.pgcode) == 'INVALID_TABLE_DEFINITION')
+
+    if force and already_exists:
+        return
+
+    raise stmt.err
+
+
+def add_primary_keys(conn_str, model_version, force=False):
+    """Add primary keys to post-transformation tables.
+
+    This will add the primary keys to an instance of the PEDSnet data model.
+
+    Make sure the search path is configured properly in the connection
+    string or in the runtime environment.
+
+    :param str conn_str:      database connection string
+    :param str model_version: the model version of the PEDSnet database
+    :param bool force: ignore benign errors if true
+    :returns:                 True if the function succeeds
+    :rtype:                   bool
+    :raises DatabaseError:    if any of the statement executions cause errors
+    """
+
+    # Log start of the function and set the starting time.
+    log_dict = combine_dicts({'model_version': model_version, 'force': force},
+                             get_conn_info_dict(conn_str))
+    task = 'adding primary keys'
+    logger.info({'msg': task}, log_dict)
```

Comment:
Also, maybe the `'msg'` should be `'starting {0}'.format(task)` for consistency? #nittynitnit